kvserver: record follower write bytes in replica load

wenyihu6 opened the pull request:
**kvserver: takes write and ingested bytes for recordRequestWriteBytes**

Previously, recordRequestWriteBytes took a *kvadmission.StoreWriteBytes, which
was unnecessary since we're only recording write and ingested bytes, not
anything admission control specific. This commit changes the function to accept
the raw byte counts directly. This also sets up a future change where follower
replicas would not pass in StoreWriteBytes during recording.

Epic: none
Release note: none

---

**kvserver: record follower write bytes in replica load**

Previously, write bytes for range load were only tracked on leaseholder replicas
during proposal, leaving write bytes written to follower replicas unaccounted
for. This commit updates the tracking to include write bytes on follower
replicas as well.

Epic: none
Release note: none

---

**kvserver: record follower write bytes during recordStatsOnCommit**

Previously, we recorded write bytes for follower replicas in
runPostAddTriggersReplicaOnly, right before apply, which could be inaccurate
since stats should be recorded at commit time. This commit moves the recording
to recordStatsOnCommit by including the write bytes in the batch stats.

Epic: none
Release note: none

---

**kvserver: rename numWriteBytes to numWriteAndIngestedBytes**

This commit renames numWriteBytes in appBatchStats to numWriteAndIngestedBytes
for accuracy, as it accounts for both bytes written to the replica and bytes
ingested via SSTables.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/150379)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: takes write and ingested bytes for recordRequestWriteBytes

Previously, recordRequestWriteBytes took a *kvadmission.StoreWriteBytes, which
was unnecessary since we're only recording write and ingested bytes, not
anything admission control specific. This commit changes the function to accept
the raw byte counts directly. This also sets up a future change where follower
replicas would not pass in StoreWriteBytes during recording.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: record follower write bytes in replica load

Previously, write bytes for range load were only tracked on leaseholder replicas
during proposal, leaving write bytes written to follower replicas unaccounted
for. This commit updates the tracking to include write bytes on follower
replicas as well.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: record follower write bytes during recordStatsOnCommit

Previously, we recorded write bytes for follower replicas in
runPostAddTriggersReplicaOnly, right before apply, which could be inaccurate
since stats should be recorded at commit time. This commit moves the recording
to recordStatsOnCommit by including the write bytes in the batch stats.

Epic: none
Release note: none

tbg reviewed (changes requested):
Looking good! No major comments, but I think there's a buglet here.

> Reviewed 1 of 1 files at r1, 4 of 4 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-:-OVRP7m6A1IVXaftdVHm:b-p1yn94)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`-- commits` line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVRP2sg8qOHwoRQVHHf:-OVRP2sg8qOHwoRQVHHg:bb7462z):*
TODO
___
*[`pkg/kv/kvserver/app_batch.go` line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVRP5WC388IIDVBLxB1:-OVRP5WC388IIDVBLxB2:br1g6mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b9e2401919ca59f0af13f1eb095b99cecc350b6/pkg/kv/kvserver/app_batch.go#L36)):*
> ```Go
> 	numEmptyEntries            int
> 	numAddSST, numAddSSTCopies int
> 	numWriteBytes              int64
> ```

// also ingested bytes
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVROD-03IHRxuWNXhez:-OVROD-03IHRxuWNXhf-:b-dj6vqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5460c6159eb01bfc93127b9766cb79de3abafeb/pkg/kv/kvserver/replica_app_batch.go#L257)):*
> ```Go
> 	// apply in a standalone context.
> 	//
> 	// TODO(irfansharif): This code block can be removed once below-raft
> ```

This TODO should move to the block related to `ApplyAdmissionControl`.
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVRNwu17Y7Dv0d_eByM:-OVRNwu17Y7Dv0d_eByN:br4fokz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5460c6159eb01bfc93127b9766cb79de3abafeb/pkg/kv/kvserver/replica_app_batch.go#L271)):*
> ```Go
> 			b.r.recordRequestWriteBytes(writeBytes, ingestedBytes)
> 		}
> 		if cmd.ApplyAdmissionControl() {
> ```

Shouldn't this be `!cmd.ApplyAdmissionControl()`? The old code was `if !local && !applyAdmissionControl`.
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 734 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVROn5m175RF4qNmKON:-OVROn5n1DehZCYi1cs_:b-f3x1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b9e2401919ca59f0af13f1eb095b99cecc350b6/pkg/kv/kvserver/replica_app_batch.go#L734)):*
> ```Go
> 	b.applyStats.numBatchesProcessed++
> 	b.applyStats.followerStoreWriteBytes.Merge(b.followerStoreWriteBytes)
> 	b.r.recordRequestWriteBytes(b.ab.numWriteBytes)
> ```

This should be functionally the same (apply is very short, so shouldn't matter), but I prefer this approach because it's more principled and potentially a little more efficient.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: rename numWriteBytes to numWriteAndIngestedBytes

This commit renames numWriteBytes in appBatchStats to numWriteAndIngestedBytes
for accuracy, as it accounts for both bytes written to the replica and bytes
ingested via SSTables.

Epic: none
Release note: none

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150379)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`-- commits` line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVRP2sg8qOHwoRQVHHf:-OVRtePH1zCEQWQ0C4xW:btrvzh7):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

TODO
</blockquote></details>

Pushed an empty commit at the end to remind me deleting it during the final rebase. 
___
*[`pkg/kv/kvserver/app_batch.go` line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVRP5WC388IIDVBLxB1:-OVRubUc0bEFsBf68vtS:b7ji28l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b9e2401919ca59f0af13f1eb095b99cecc350b6/pkg/kv/kvserver/app_batch.go#L36)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

// also ingested bytes
</blockquote></details>

I renamed this to numWriteAndIngestedBytes. 
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 257 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVROD-03IHRxuWNXhez:-OVRv6n11ctqLK1SrE4y:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5460c6159eb01bfc93127b9766cb79de3abafeb/pkg/kv/kvserver/replica_app_batch.go#L257)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This TODO should move to the block related to `ApplyAdmissionControl`.
</blockquote></details>

Done. 
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVRNwu17Y7Dv0d_eByM:-OVRv8T793h_DcKUuovp:b-bpmkhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5460c6159eb01bfc93127b9766cb79de3abafeb/pkg/kv/kvserver/replica_app_batch.go#L271)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Shouldn't this be `!cmd.ApplyAdmissionControl()`? The old code was `if !local && !applyAdmissionControl`.
</blockquote></details>

:grimacing: good catch!
___
*[`pkg/kv/kvserver/replica_app_batch.go` line 734 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150379#-OVROn5m175RF4qNmKON:-OVRvT4D-a2VzIBL4xki:bg13jod) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b9e2401919ca59f0af13f1eb095b99cecc350b6/pkg/kv/kvserver/replica_app_batch.go#L734)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This should be functionally the same (apply is very short, so shouldn't matter), but I prefer this approach because it's more principled and potentially a little more efficient.
</blockquote></details>

Ack. Removed the TODO. 


<!-- Sent from Reviewable.io -->
